counter/src/main/java/org/infinispan/counter/impl/manager/CacheHolder.java

```diff
@@ -56,15 +59,16 @@ void registerNotificationManager(CounterManagerNotificationManager notificationM
       notificationManager.listenOn(counterCache);
    }
 
-   boolean addConfiguration(String name, CounterConfiguration configuration) {
-      CounterConfiguration existing = checkAndStoreConfiguredCounterConfiguration(name);
-      if (existing != null) {
-         return false;
-      }
-      Cache<String, CounterConfiguration> cache = configuration.storage() == Storage.VOLATILE ?
-            configurationCache.withFlags(Flag.SKIP_CACHE_STORE, Flag.SKIP_CACHE_LOAD) :
-            configurationCache;
-      return cache.putIfAbsent(name, configuration) == null;
+   CompletableFuture<Boolean> addConfigurationAsync(String name, CounterConfiguration configuration) {
+      return checkAndStoreConfiguredCounterConfigurationAsync(name)
+            .thenCompose(fileConfig -> {
+               if (fileConfig != null) {
+                  //counter configuration exists in configuration file!
+                  return CompletableFuture.completedFuture(fileConfig);
+               }
+               //put configuration in cache
+               return configCacheWithFlags(configuration).putIfAbsentAsync(name, configuration);
```

Comment:
I don't follow; aren't you calling the `putIfAbsentAsync` in `checkAndStoreConfiguredCounterConfigurationAsync` already?

Reply:
`checkAndStoreConfiguredCounterConfigurationAsync` only checks the counters defined in the configuration file. if the counter exists, I put it in the cache (`putIfAbsent` in the method)
if it is not defined in the configuration, we put the in the cache with the user's `CounterConfiguration`.

Reply:
no it is lazy since day 2 :) the configuration is added to the cache when they are accessed for the first time.